File: src/test/java/de/iteratec/iteraOfficeMap/WorkplaceServiceTest.java
package de.iteratec.iteraOfficeMap;

import de.iteratec.iteraOfficeMap.exceptions.AlreadyExistsException;
import de.iteratec.iteraOfficeMap.exceptions.DoesNotExistException;
import de.iteratec.iteraOfficeMap.utility.DateUtility;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.util.Date;
import java.util.List;

import static java.util.Arrays.asList;
import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest
@ExtendWith(SpringExtension.class)
public class WorkplaceServiceTest {

    @Autowired
    private ReservationRepository reservationRepository;

    @Autowired
    private WorkplaceRepository workplaceRepository;

    @Autowired
    private WorkplaceService workplaceService;

    private Workplace workplace1;
    private Workplace workplace2;
    private Workplace workplace3;
    private Workplace workplace4;
    private Long currentDay = DateUtility.startOfDay(new Date());

    private void createAndSaveWorkplaces() {
        workplace1 = new Workplace("First", 1, 1, "mapId");
        workplace2 = new Workplace("Second", 2, 2, "mapId");
        workplace3 = new Workplace("Third", 3, 3, "mapId");
        workplace4 = new Workplace("Fourth", 4, 4, "mapId");
        workplaceRepository.saveAll(asList(workplace1, workplace2, workplace3));
    }

    void assertWorkplaceEqualsWorkplaceDTO(Workplace workplace, WorkplaceDTO workplaceDTO) {

        // assertEquals(workplace.getId(),workplaceDTO.getId());
        assertEquals(workplace.getName(), workplaceDTO.getName());
        assertEquals(workplace.getX(), workplaceDTO.getX());
        assertEquals(workplace.getY(), workplaceDTO.getY());
    }

    @BeforeEach
    void init() {
        reservationRepository.deleteAll(reservationRepository.findAll());
        workplaceRepository.deleteAll(workplaceRepository.findAll());
    }


    @Test
    void getAllWorkplacesTest() {
        createAndSaveWorkplaces();
        List<WorkplaceDTO> workplaceDTOList = workplaceService.getAllWorkplaces();

        assertEquals(3, workplaceDTOList.size());
        assertWorkplaceEqualsWorkplaceDTO(workplace1, workplaceDTOList.get(0));
        assertWorkplaceEqualsWorkplaceDTO(workplace2, workplaceDTOList.get(1));
        assertWorkplaceEqualsWorkplaceDTO(workplace3, workplaceDTOList.get(2));
    }

    @Test
    void getAllWorkplacesEmptyListTest() {
        assertEquals(0, workplaceService.getAllWorkplaces().size());
    }

    @Test
    void addWorkPlaceTest() {
        workplace4 = new Workplace("Fourth", 4, 4, "mapId");
        AddWorkplaceDTO addWorkplaceDTO = new AddWorkplaceDTO(workplace4);
        workplaceService.addNewWorkplace(addWorkplaceDTO);
        List<WorkplaceDTO> workplaceDTOList = workplaceService.getAllWorkplaces();

        assertEquals(1, workplaceDTOList.size());
        assertWorkplaceEqualsWorkplaceDTO(workplace4, workplaceDTOList.get(0));
    }

    @Test
    void getStatusOfFreeWorkplaceTest() {
        workplace4 = new Workplace("fourth", 4, 4, "mapId");
        workplaceRepository.save(workplace4);

        assertEquals(Status.FREE, workplaceService.getStatus(workplace4.getId()));
    }

    @Test
    void getStatusOfReservedTest() {
        workplace4 = new Workplace("fourth", 4, 4, "mapId");
        workplaceRepository.save(workplace4);
        Reservation reserveToday = new Reservation(currentDay, workplace4, "Sascha");
        reservationRepository.save(reserveToday);

        assertEquals(Status.NORMALRESERVATION, workplaceService.getStatus(workplace4.getId()));
    }

    @Test
    void addAlreadyExistingWorkPlaceWithSameCoordinates() {
        Assertions.assertThrows(AlreadyExistsException.class, () -> {
            workplace4 = new Workplace("fourth", 4, 4, "mapId");
            workplace3 = new Workplace("third", 4, 4, "mapId");
            AddWorkplaceDTO addWorkplaceDTO = new AddWorkplaceDTO(workplace4);
            AddWorkplaceDTO addWorkplaceDTONew = new AddWorkplaceDTO(workplace3);
            workplaceService.addNewWorkplace(addWorkplaceDTO);
            workplaceService.addNewWorkplace(addWorkplaceDTONew);
        });
    }

    @Test
    void addAlreadyExistingWorkPlaceWithSameName() {
        Assertions.assertThrows(AlreadyExistsException.class, () -> {
            workplace4 = new Workplace("fourth", 4, 4, "mapId");
            workplace3 = new Workplace("fourth", 3, 3, "mapId");
            AddWorkplaceDTO addWorkplaceDTO = new AddWorkplaceDTO(workplace4);
            AddWorkplaceDTO addWorkplaceDTONew = new AddWorkplaceDTO(workplace3);
            workplaceService.addNewWorkplace(addWorkplaceDTO);
            workplaceService.addNewWorkplace(addWorkplaceDTONew);
        });
    }

    @Test
    void deleteWorkPlaceTest() {
        createAndSaveWorkplaces();
        DeleteWorkplaceDTO deleteWorkplaceDTO = new DeleteWorkplaceDTO(workplace1.getId());
        workplaceService.deleteWorkplace(deleteWorkplaceDTO);
        List<Workplace> workplaceList = workplaceRepository.findAll();
        assertEquals(2, workplaceList.size());
        assertEquals(workplace2, workplaceList.get(0));
        assertEquals(workplace3, workplaceList.get(1));
    }

    @Test
    void DeleteNonExistingWorkplaceTest() {
        Assertions.assertThrows(DoesNotExistException.class, () -> {
            createAndSaveWorkplaces();
            workplaceRepository.deleteById(workplace1.getId());
            DeleteWorkplaceDTO deleteWorkplaceDTO = new DeleteWorkplaceDTO(workplace1.getId());
            workplaceService.deleteWorkplace(deleteWorkplaceDTO);
        });
    }
}

File: src/test/java/de/iteratec/officemap/workplace/WorkplaceServiceTest.kt
package de.iteratec.officemap.workplace

import de.iteratec.officemap.exceptions.AlreadyExistsException
import de.iteratec.officemap.exceptions.DoesNotExistException
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.containsInAnyOrder
import org.hamcrest.Matchers.empty
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit.jupiter.SpringExtension

@SpringBootTest
@ActiveProfiles("dev")
@ExtendWith(SpringExtension::class)
class WorkplaceServiceTest
@Autowired constructor(
        private val workplaceRepository: WorkplaceRepository,
        private val workplaceService: WorkplaceService
) {

    private val workplace1 = Workplace(null, "First", 1, 1, "mapId", "Zwei Bildschirme")
    private val workplace2 = Workplace(null, "Second", 2, 2, "mapId", "Zwei Bildschirme")
    private val workplace3 = Workplace(null, "Third", 3, 3, "mapId", "Zwei Bildschirme")
    private val workplace4 = Workplace(null, "Fourth", 4, 4, "mapId", "Zwei Bildschirme")

    @BeforeEach
    fun deleteAllWorkplaces() {
        workplaceRepository.deleteAll(workplaceRepository.findAll())
    }

    @Test
    fun `when no workplaces exist none can be found`() {
        assertThat(workplaceService.getAllWorkplaces(), empty())
    }

    @Test
    fun `when one workplace exists it can be found`() {
        workplaceRepository.save(workplace1)
        assertThat(workplaceService.getAllWorkplaces(), containsInAnyOrder(workplace1))
    }

    @Test
    fun `when 4 workplaces exist all can be found`() {
        workplaceRepository.saveAll(listOf(workplace1, workplace2, workplace3, workplace4))
        assertThat(workplaceService.getAllWorkplaces(), containsInAnyOrder(workplace1, workplace2, workplace3, workplace4))
    }

    @Test
    fun `add 1 workplace`() {
        assertThat(workplaceService.getAllWorkplaces(), empty())
        workplaceService.addNewWorkplace(workplace1)
        assertThat(workplaceService.getAllWorkplaces(), containsInAnyOrder(workplace1))
    }

    @Test
    fun `add second workplace at same coordinates throws AlreadyExistsException`() {
        workplaceService.addNewWorkplace(workplace1)
        assertThrows(AlreadyExistsException::class.java) {
            workplaceService.addNewWorkplace(workplace1.copy(name = "otherName"))
        }
    }

    @Test
    fun `add second workplace with same name but different coordinates throws AlreadyExistsException`() {
        workplaceService.addNewWorkplace(workplace1)
        assertThrows(AlreadyExistsException::class.java) {
            workplaceService.addNewWorkplace(workplace1.copy(x = 123, y = 123))
        }
    }

    @Test
    fun `delete last workplace`() {
        val workplace = workplaceRepository.save(workplace1)
        val id = requireNotNull(workplace.id)

        assertThat(workplaceService.getAllWorkplaces(), containsInAnyOrder(workplace))
        workplaceService.deleteWorkplace(id)
        assertThat(workplaceService.getAllWorkplaces(), empty())
    }

    @Test
    fun `delete one of two workplaces`() {
        val workplaces = workplaceRepository.saveAll(listOf(workplace1, workplace2))
        val (first, second) = workplaces

        val firstId = requireNotNull(first.id)

        assertThat(workplaceService.getAllWorkplaces(), containsInAnyOrder(first, second))
        workplaceService.deleteWorkplace(firstId)
        assertThat(workplaceService.getAllWorkplaces(), containsInAnyOrder(second))
    }

    @Test
    fun `delete a non-existing workplace throws a DoesNotExistException`() {
        assertThrows(DoesNotExistException::class.java) {
            workplaceService.deleteWorkplace(1234)
        }
    }

}
